Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateServiceImpl.java

```diff
@@ -0,0 +1,163 @@
+        for (Map<String, Object> item : array) {
+            Object keyValue = item.get(key.get("groupKey"));
+            if (!map.containsKey(keyValue)) {
+                Map<String, Object> group = new HashMap<>();
+                group.put(key.get("groupKey"), keyValue);
+                group.put(key.get("childrenKey"), new ArrayList<Map<String, Object>>());
+                map.put(keyValue, group);
+            }
+            ((List)map.get(keyValue).get(key.get("childrenKey"))).add(item);
+        }
+        List<Map<String, Object>> groups = new ArrayList<>(map.values());
+
+        // find common properties from the children and copy them to the group
+        for (Map<String, Object> group : groups) {
+            Map<String, Object> common = findCommonProperties((List<Map<String, Object>>) group.get(key.get("childrenKey")));
+            group.putAll(common);
+        }
+
+        return groups;
+    }
+
+    /**
+     * Based on the provided data from the source file, generates a hierarchical data structure for the template to process.
+     * @param data data to be mapped for the template configuration Json.
+     * @param keys key-value structure of values in which the data is going to be mapped.
+     * @return List<Map<String, Object>> List of hierarchical structure map based on provided keys for the template processor.
+     */
+    private List<Map<String, Object>> groupByKeys(List<Map<String, Object>> data, List<Map<String, String>> keys) {
+        if (keys.isEmpty()) {
+            return data;
+        }
+        Map<String, String> firstKey = keys.get(0);
+        List<Map<String, String>> nextKeys = keys.subList(1, keys.size());
+        List<Map<String, Object>> groups = groupByKey(data, firstKey);
+        for (Map<String, Object> group : groups) {
+            group.put(firstKey.get("childrenKey"), groupByKeys((List<Map<String, Object>>) group.get(firstKey.get("childrenKey")), nextKeys));
+        }
+        return groups;
+    }
+
+    /**
+     * find common properties from each hierarchical section or subgroup.
+     *
+     * @param data list of sub data from the mapped group structure.
+     * @return Map<String, Object> generated from common properties sor each section of the group structure.
+     */
+    private Map<String, Object> findCommonProperties(List<Map<String, Object>> data) {
+        if (data.isEmpty()) {
+            return new HashMap<>();
+        }
+        Map<String, Object> common = new HashMap<>(data.get(0));
+        for (Map<String, Object> item : data) {
+            for (Map.Entry<String, Object> entry : new HashMap<>(common).entrySet()) {
+                String k = entry.getKey();
+                Object v = entry.getValue();
+                if (!item.get(k).equals(v)) {
+                    common.remove(k);
+                }
+            }
+        }
+        return common;
+    }
+
+    private List<Map<String, Object>> readCSV(CSVReader reader) throws IOException {
```

Comment:
This would always return a `List<Map<String,String>>` right.  I don't see why you are using an Object here.  Unless I'm missing something that should be changed.

Reply:
We are using this List structure List<Map<String, Object>> for all the hierarchy within the map, if I change this to <String, String> I would need to add extra steps for this to be converted to obj as initial data source obj and then send it to the groupByKeys(..) method